gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -392,14 +395,30 @@ private void addJars(Path jarFileDir, String jarFileList, Configuration conf)
     for (String jarFile : SPLITTER.split(jarFileList)) {
       Path srcJarFile = new Path(jarFile);
       FileStatus[] fileStatusList = lfs.globStatus(srcJarFile);
+
+      jarFileLoop:
       for (FileStatus status : fileStatusList) {
-        // SNAPSHOT jars should not be shared, as different jobs may be using different versions of it
-        Path baseDir = status.getPath().getName().contains("SNAPSHOT") ? this.unsharedJarsDir : jarFileDir;
-        // DistributedCache requires absolute path, so we need to use makeQualified.
-        Path destJarFile = new Path(this.fs.makeQualified(baseDir), status.getPath().getName());
-        if (!this.fs.exists(destJarFile)) {
-          // Copy the jar file from local file system to HDFS
-          this.fs.copyFromLocalFile(status.getPath(), destJarFile);
+        // For each FileStatus there are chances it could fail in copying at the first attempt, due to file-existence
+        // or file-copy is ongoing by other job instance since all Gobblin jobs share the same jar file directory.
+        // the retryCount is to avoid cases (if any) where retry is going too far and causes job hanging.
+        int retryCount = 0 ;
+        Path destJarFile = calculateDestJarFile(status, jarFileDir);
+          // Adding destJarFile into HDFS until it exists and the size of file on targetPath matches the one on local path.
+        while(!this.fs.exists(destJarFile) || fs.listStatus(destJarFile).length != status.getLen()) {
```

Comment:
In the file exists, but size does not match case we need to give time for the other process to copy the file.